threedi_models_and_simulations/widgets/simulation_wizard.py

```diff
@@ -681,6 +696,66 @@ def connect_signals(self):
         self.pb_upload_2d_laterals.clicked.connect(partial(self.load_csv, self.TYPE_2D))
         self.cb_2d_interpolate.stateChanged.connect(partial(self.interpolate_changed, self.TYPE_2D))
 
+    def setup_substance_concentrations(self):
+        if hasattr(self, "groupbox"):
+            self.groupbox.setParent(None)
+        if not self.substances:
+            return
+        substance_concentration_widget = SubstanceConcentrationsWidget(self.substances, self.handle_substance_timesteps)
+        self.groupbox = substance_concentration_widget.groupbox
+        self.substance_concentrations = substance_concentration_widget.substance_concentrations
+        parent_layout = self.layout()
+        parent_layout.addWidget(self.groupbox, 3, 2)
+
+    def handle_substance_header(self, substance_list):
+        """
+        Fetch first csv row and handle potential header.
+        Return None if fetch successful or error message if file is empty or have invalid structure.
+        """
+        error_message = None
+        if not substance_list:
+            error_message = "Substance concentration list is empty!"
+            return error_message
+        header = substance_list[0]
+        if len(header) != 2:
+            error_message = "Wrong csv format for substance concentrations!"
+        if error_message is None:
+            try:
+                timeseries_candidate = header[-1]
+                [[float(f) for f in line.split(",")] for line in timeseries_candidate.split("\n")]
+            except ValueError:
+                substance_list.pop(0)
+        return error_message
+
+    def handle_substance_timesteps(self, substance_list):
+        """
+        First, check if lateral values are uploaded.
+        Second, check if substance concentrations timesteps match exactly the lateral values timesteps.
+        Return None if they match or error message if not.
+        """
+        error_message = self.handle_substance_header(substance_list)
+        if not self.laterals_1d_timeseries and not self.laterals_2d_timeseries:
+            error_message = "No laterals uploaded yet!"
+        if not substance_list:
+            error_message = "Substance concentrations list is empty!"
+        if error_message is None:
+            laterals_timeseries = {**self.laterals_1d_timeseries, **self.laterals_2d_timeseries}
+            for lat_id, timeseries in substance_list:
+                lateral = laterals_timeseries.get(lat_id)
+                if lateral is None:
+                    error_message = f"Laterals with ID {lat_id} not found!"
+                    break
+                lateralValues = lateral["values"]
+                laterals_timesteps = [t for (t, _) in lateralValues]
+                concentrations = [[float(f) for f in line.split(",")] for line in timeseries.split("\n")]
```

Comment:
Same logic repeats in a line 725. Please move the logic to separate static method or utils function.